stratum/hal/bin/bmv2/README.md

```diff
@@ -69,5 +82,136 @@ cp stratum/hal/bin/bmv2/update_config.py /tmp/ && \
     --grpc-addr <YOUR_HOST_IP_ADDRESS>:9559 --json <prog>.json --p4info <prog>.proto.txt
 ```
 
+<!--
+FIXME(bocon): Update to use the P4RT shell
+-->
+
 You can use the loopback program under `stratum/pipelines/loopback/p4c-out/bmv2`
 if you do not have your own P4 program.
+
+## Developer builds of `stratum_bmv2`
+
+The following steps are for developers who want to do fast, incremental builds of
+`stratum_bmv2` and then run them in the development environment.
+
+### Setup
+
+Start your development environment container with priveledged mode in the host
+network namespace. This will allow you to create new virtual network interfaces
+(veth pairs) and interact with them.
+
+```bash
+./setup_dev_env.sh -- --privileged --network=host
+```
+
+Next, create some virtual network interfaces (veth pairs) for `stratum_bmv2`.
+
+```bash
+sudo veth_setup.sh
+```
+
+When you are finished, you can remove the interfaces.
+```bash
+sudo veth_teardown.sh
+```
+
+If you don't have the `veth_setup.sh` command installed, you can
+add interfaces manually:
+```bash
+sudo ip link add name veth0 type veth peer name veth1
+sudo ip link add name veth2 type veth peer name veth3
+sudo ip link set dev veth0 up
+sudo ip link set dev veth1 up
+sudo ip link set dev veth2 up
+sudo ip link set dev veth3 up
+```
+
+By default, Stratum will bind `veth0` as port 1 and `veth2` as port 2.
+More generally, Stratum will bind port `X` to `veth[(X-1) * 2]`.
+
+You can send and receive packets to/from port 1 using `veth1` and
+to/from port 2 using `veth3`. More generally, you can send and receive
+packets for Stratum port `X` using `veth[(X-1) * 2 + 1]`.
+
+This port mapping is a little confusing, but Statum just follows the
+convention from BMv2. If you choose to use a different naming scheme,
+be sure to update the port `name` in the chassis config.
+
+### Building and running `stratum_bmv2` for development
+
+You can build and run the Stratum binary directly using the following
+Bazel command:
+
+```bash
+export BMV2_DIR=$(bazel info workspace)/stratum/hal/bin/bmv2
+bazel build //stratum/hal/bin/bmv2:stratum_bmv2
+sudo bazel-bin/stratum/hal/bin/bmv2/stratum_bmv2 \
+    -persistent_config_dir=/tmp/ \
+    -chassis_config_file=${BMV2_DIR}/chassis_config.pb.txt \
+    -initial_pipeline=${BMV2_DIR}/dummy.json \
+    -forwarding_pipeline_configs_file=/tmp/bmv2_pipeline_cfg \
+    -bmv2_log_level=debug
+```
+
```

Comment:
```
I20210802 18:56:46.491569   408 p4_service.cc:121] Pushing the saved forwarding pipeline configs read from /tmp/bmv2_pipeline_cfg...
E20210802 18:56:46.491609   408 utils.cc:112] StratumErrorSpace::ERR_FILE_NOT_FOUND: /tmp/bmv2_pipeline_cfg not found.
E20210802 18:56:46.491842   408 utils.cc:68] Return Error: ReadFileToString(filename, &text) failed with StratumErrorSpace::ERR_FILE_NOT_FOUND: /tmp/bmv2_pipeline_cfg not found.
W20210802 18:56:46.491852   408 p4_service.cc:130] No saved forwarding pipeline config found at /tmp/bmv2_pipeline_cfg. This is normal when the switch is just installed and no master controller is connected yet. 
``` 
I assume this is expected - is there a canned config for bmv2 that can be saved to /tmp ? or pushed for that matter?

Reply:
I think fabric-tna should be pushable to bmv2: https://github.com/stratum/fabric-tna